### PR TITLE
use ty::Binder in rustdoc instead of `skip_binder`

### PR DESCRIPTION
**Pull request by oli-obk**
r? @GuillaumeGomez 

this is a preliminary cleanup required to be able to normalize correctly/conveniently in rustdoc

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1459,8 +1460,11 @@ fn clean_qpath<'tcx>(hir_ty: &hir::Ty<'tcx>, cx: &mut DocContext<'tcx>) -> Type
...
             // Try to normalize `<X as Y>::T` to a type
             let ty = hir_ty_to_ty(cx.tcx, hir_ty);
-            if let Some(normalized_value) = normalize(cx, ty::Binder::dummy(ty)) {
-                return clean_middle_ty(normalized_value, cx, None);
+            // `hir_to_ty` can return projection types with escaping vars for GATs, e.g. `<() as Trait>::Gat<'_>`
+            if !ty.has_escaping_bound_vars() {
+                if let Some(normalized_value) = normalize(cx, ty::Binder::dummy(ty)) {
+                    return clean_middle_ty(normalized_value, cx, None);
+                }
             }
```
Comment (oli-obk) [inline src/librustdoc/clean/mod.rs:1468]: This is necessary due to the case that troubled us previously in https://github.com/rust-lang/rust/issues/94683

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! I don't think it'll have a perf impact but I'll let it up to you whether or not it should be in a rollup. r=me once CI pass.

**Comment by oli-obk** [comment]
Even if it regresses perf for rustdoc, we should still land it as it's a correctness issue and will prevent ICEs down the road

**Comment by GuillaumeGomez** [comment]
It's very acceptable. Let's go then!

@bors r+ rollup=never

**Comment by aDotInTheVoid** [comment]
> we should still land it as it's a correctness issue 

May I ask, if this is a correctness fix, is their anything in particular that this fixes, or is this just moving to a more principled way that's less brittle.

**Comment by oli-obk** [comment]
The latter. Right now all uses of normalization in rustdoc need to guard against broken types. This fixes that. I'll follow up with cleanups, but those may change rustdoc behaviour. The changes will be improvements to generics-heavy code

**Comment by oli-obk** [comment]
The regressions are expected, solely in rustdoc and part of a bugfix

**Pull request merged**